FirebasePerformance/Sources/FPRNanoPbUtils.m

```diff
@@ -0,0 +1,452 @@
+
+#pragma mark - Private methods
+
+pb_bytes_array_t *FPREncodeData(NSData *data) {
+  pb_bytes_array_t *pbBytesArray = calloc(1, PB_BYTES_ARRAY_T_ALLOCSIZE(data.length));
+  if (pbBytesArray != NULL) {
+    [data getBytes:pbBytesArray->bytes length:data.length];
+    pbBytesArray->size = (pb_size_t)data.length;
+  }
+  return pbBytesArray;
+}
+
+pb_bytes_array_t *FPREncodeString(NSString *string) {
+  NSData *stringBytes = [string dataUsingEncoding:NSUTF8StringEncoding];
+  return FPREncodeData(stringBytes);
+}
+
+NSData *FPRDecodeData(pb_bytes_array_t *pbData) {
+  NSData *data = [NSData dataWithBytes:&(pbData->bytes) length:pbData->size];
+  return data;
+}
+
+NSString *FPRDecodeString(pb_bytes_array_t *pbData) {
+  NSData *data = FPRDecodeData(pbData);
+  return [NSString stringWithCString:[data bytes] encoding:NSUTF8StringEncoding];
+}
+
+NSMutableDictionary<NSString *, NSString *> *FPRDecodeCustomAttributes(
+    struct _firebase_perf_v1_ApplicationInfo_CustomAttributesEntry *customAttributes,
+    NSInteger count) {
+  NSMutableDictionary<NSString *, NSString *> *dict = [NSMutableDictionary dictionary];
+  for (int i = 0; i < count; i++) {
+    NSString *key = FPRDecodeString(customAttributes[i].key);
+    NSString *value = FPRDecodeString(customAttributes[i].value);
+    dict[key] = value;
+  }
+  return dict;
+}
+
+#pragma mark - Public methods
+
+firebase_perf_v1_PerfMetric GetPerfMetricMessage(NSString *appID) {
+  firebase_perf_v1_PerfMetric perfMetricMessage = firebase_perf_v1_PerfMetric_init_default;
+  perfMetricMessage.application_info = GetApplicationInfoMessage();
+  perfMetricMessage.application_info.google_app_id = FPREncodeString(appID);
+
+  return perfMetricMessage;
+}
+
+firebase_perf_v1_ApplicationInfo GetApplicationInfoMessage() {
+  firebase_perf_v1_ApplicationInfo appInfoMessage = firebase_perf_v1_ApplicationInfo_init_default;
+  firebase_perf_v1_IosApplicationInfo iosAppInfo = firebase_perf_v1_IosApplicationInfo_init_default;
+  NSBundle *mainBundle = [NSBundle mainBundle];
+  iosAppInfo.bundle_short_version =
+      FPREncodeString([mainBundle infoDictionary][@"CFBundleShortVersionString"]);
+  iosAppInfo.sdk_version = FPREncodeString([NSString stringWithUTF8String:kFPRSDKVersion]);
+  iosAppInfo.network_connection_info.network_type = FPRNetworkConnectionInfoNetworkType();
+#ifdef TARGET_HAS_MOBILE_CONNECTIVITY
+  CTTelephonyNetworkInfo *networkInfo = NetworkInfo();
+  CTCarrier *provider = networkInfo.subscriberCellularProvider;
+  NSString *mccMnc = FPRValidatedMccMnc(provider.mobileCountryCode, provider.mobileNetworkCode);
+  if (mccMnc) {
+    iosAppInfo.mcc_mnc = FPREncodeString(mccMnc);
+  }
+  if (iosAppInfo.network_connection_info.network_type ==
+      firebase_perf_v1_NetworkConnectionInfo_NetworkType_MOBILE) {
+    iosAppInfo.network_connection_info.mobile_subtype = FPRCellularNetworkType();
+  }
+#endif
+  appInfoMessage.ios_app_info = iosAppInfo;
+
+  NSDictionary<NSString *, NSString *> *attributes =
```

Comment:
Similar to decodeCustomAttributes, can you create a method as encodeCustomAttributes?

Or even better if we could generalize that as "encodeMap" and "decodeMap". WDYT?

Reply:
Sounds good! Create encode and decode for stringToStringMap and stringToIntMap